src/hotspot/share/opto/vector.cpp

```diff
@@ -503,34 +509,38 @@ Node* PhaseVector::get_loaded_payload(VectorUnboxNode* vec_unbox) {
    return NULL;
 }
 
-void PhaseVector::expand_vunbox_node_mf(VectorUnboxNode* vec_unbox) {
+void PhaseVector::expand_vunbox_node_vector(VectorUnboxNode* vec_unbox) {
   if (vec_unbox->outcnt() > 0) {
     GraphKit kit;
     PhaseGVN& gvn = kit.gvn();
 
-    Node* obj = vec_unbox->obj();
-    const TypeInstPtr* tinst = gvn.type(obj)->isa_instptr();
-    ciInstanceKlass* from_kls = tinst->instance_klass();
-    const TypeVect* vt = vec_unbox->bottom_type()->is_vect();
-    BasicType bt = vt->element_basic_type();
-    BasicType masktype = bt;
-    int num_elem = vt->length();
-
-    int elem_size = type2aelembytes(bt);
     Node* vec_val_load = get_loaded_payload(vec_unbox);
+
     if (vec_val_load == NULL) {
-      assert(obj->isa_InlineType(), "");
-      ciSymbol* payload_sig = ciSymbol::make(VectorSupport::get_vector_payload_field_signature(bt, num_elem)->as_C_string());
-      ciSymbol* payload_name = ciSymbol::make(vmSymbols::payload_name()->as_C_string());
-      ciField* payload = from_kls->get_field_by_name(payload_name, payload_sig, false);
+      Node* obj = vec_unbox->obj();
+      const TypeInstPtr* tinst = gvn.type(obj)->isa_instptr();
+      assert(VectorSupport::is_vector(tinst->instance_klass()), "");
+      ciInlineKlass* box_klass = static_cast<ciInlineKlass*>(tinst->instance_klass());
+
+      const TypeVect* vt = vec_unbox->bottom_type()->is_vect();
+      BasicType bt = vt->element_basic_type();
+      int num_elem = vt->length();
+
+      assert(VectorSupport::is_vector(box_klass), "");
+      ciField* payload_field = box_klass->declared_nonstatic_field_at(0);
+      int offset = payload_field->offset();
+      if (!payload_field->is_flattened()) {
+        ciInlineKlass* payload_klass = static_cast<ciInlineKlass*>(payload_field->type());
+        assert(VectorSupport::is_vector_payload_mf(payload_klass), "");
+        ciField* mutifield = payload_klass->declared_nonstatic_field_at(0);
+        offset += mutifield->offset();
+      }
```

Comment:
Same with expanding vbox, if the `payload_field` is not flattened, we have to load the payload instance first, decode and then load the vector mf?